src/Components/PasswordInput.test.js

```diff
@@ -0,0 +1,59 @@
+import { render, screen } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+import '@testing-library/jest-dom';
+import PasswordInput from './PasswordInput';
+
+const mockedSetPassword = jest.fn();
+
+describe('PasswordInput', () => {
+  test('renders correctly', () => {
+    render(<PasswordInput password="" setPassword={mockedSetPassword} />);
+    const hiddenPasswordInput = screen.queryByLabelText(/password:/i);
+    expect(hiddenPasswordInput).toBeInTheDocument();
+    expect(hiddenPasswordInput).toHaveTextContent('');
+
+    const visiblePasswordInput = screen.queryByRole('textbox', {
+      name: /password:/i,
+    });
+    expect(visiblePasswordInput).not.toBeInTheDocument();
+
+    const button = screen.getByRole('button');
+    const showIcon = screen.getByTitle('Show password');
+
+    expect(button).toBeInTheDocument();
+    expect(showIcon).toBeInTheDocument();
+  });
+
+  test('should change value based on password prop', () => {
+    render(
+      <PasswordInput password="ABCabc123!" setPassword={mockedSetPassword} />,
+    );
+    const inputElement = screen.getByPlaceholderText('e.g., ••••••••');
+    expect(inputElement.value).toBe('ABCabc123!');
+    expect(inputElement).toBeInTheDocument();
+  });
+
+  test('should change icon and type when button clicked', async () => {
+    const user = userEvent.setup();
+    render(<PasswordInput password="" setPassword={mockedSetPassword} />);
+    const togglePasswordVisibilityButton = screen.getByRole('button');
+
+    const passwordInput = screen.queryByLabelText(/password:/i);
+    expect(passwordInput).toBeInTheDocument();
+    expect(passwordInput.type).toBe('password');
+
+    let hideIcon = screen.queryByTestId('hide');
+    expect(hideIcon).not.toBeInTheDocument();
+
+    let showIcon = screen.queryByTestId('show');
+    expect(showIcon).toBeInTheDocument();
+
+    await user.click(togglePasswordVisibilityButton);
+
+    expect(passwordInput.type).toBe('text');
+    hideIcon = screen.queryByTestId('hide');
+    expect(hideIcon).toBeInTheDocument();
+    showIcon = screen.queryByTestId('show');
+    expect(showIcon).not.toBeInTheDocument();
```

Comment:
While it's possible to use ids it's not the preferred way of querying elements. The reason is: the user doesn't see ids. The user sees text, perceives roles, clicks on things, but the underlying implementation is invisible to them. Conversely, you may have overlapping elements where one elements becomes invisible to the user. In that case, you can't see or access the invisible element. That's why tests are so powerful: they mock the perception and behavior of the user _instead of relying on hidden implementation details only accessible to the developer_.

Reply:
Thank you very much for your review.
Since I am looking for an icon, the only alternative I have is the following:

`container.querySelector('#root > div > form > div:nth-child(2) > div > button > svg > path.`

This is even still in the rankings under ByTestId.

For the icons, I use the **@heroicons/react** package, so the implementation details are (sort of) hidden from me.

I use the **Testing Playground** Chrome plugin and it didn't show me a simpler variant.